kong/runloop/plugins_iterator.lua

```diff
@@ -85,129 +85,139 @@ local function load_plugin_configuration(route_id,
+    len = len + 1
+    threads[len] = ngx_thread_spawn(load_plugin_configuration, plugin_name,
+                                    nil, service_id, consumer_id, nil)
+  end
 
-    local plugin_configuration
+  if api_id and consumer_id then
+    len = len + 1
+    threads[len] = ngx_thread_spawn(load_plugin_configuration, plugin_name,
+                                    nil, nil, consumer_id, api_id)
+  end
 
-    repeat
+  if route_id and service_id then
+    len = len + 1
+    threads[len] = ngx_thread_spawn(load_plugin_configuration, plugin_name,
+                                    route_id, service_id, nil, nil)
+  end
 
-      if route_id and service_id and consumer_id then
-        plugin_configuration = load_plugin_configuration(route_id, service_id, consumer_id, plugin_name, nil)
-        if plugin_configuration then
-          break
-        end
-      end
+  if consumer_id then
+    len = len + 1
+    threads[len] = ngx_thread_spawn(load_plugin_configuration, plugin_name,
+                                    nil, nil, consumer_id, nil)
+  end
 
-      if route_id and consumer_id then
-        plugin_configuration = load_plugin_configuration(route_id, nil, consumer_id, plugin_name, nil)
-        if plugin_configuration then
-          break
-        end
-      end
+  if route_id then
+    len = len + 1
+    threads[len] = ngx_thread_spawn(load_plugin_configuration, plugin_name,
+                                    route_id, nil, nil, nil)
+  end
 
-      if service_id and consumer_id then
-        plugin_configuration = load_plugin_configuration(nil, service_id, consumer_id, plugin_name, nil)
-        if plugin_configuration then
-          break
-        end
-      end
+  if service_id then
+    len = len + 1
+    threads[len] = ngx_thread_spawn(load_plugin_configuration, plugin_name,
+                                    nil, service_id, nil, nil)
+  end
 
-      if api_id and consumer_id then
-        plugin_configuration = load_plugin_configuration(nil, nil, consumer_id, plugin_name, api_id)
-        if plugin_configuration then
-          break
-        end
-      end
+  if api_id then
+    len = len + 1
+    threads[len] = ngx_thread_spawn(load_plugin_configuration, plugin_name,
+                                    nil, nil, nil, api_id)
+  end
 
-      if route_id and service_id then
-        plugin_configuration = load_plugin_configuration(route_id, service_id, nil, plugin_name, nil)
-        if plugin_configuration then
-          break
-        end
-      end
+  for i = 1, len do
+    local _, cfg, err = assert(ngx_thread_wait(threads[i]))
+    if err then
+      ngx.ctx.delay_response = false
+      return responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
+    end
 
-      if consumer_id then
-        plugin_configuration = load_plugin_configuration(nil, nil, consumer_id, plugin_name, nil)
-        if plugin_configuration then
-          break
-        end
-      end
+    if cfg then
+      return cfg
+    end
```

Comment:
Seems like this could just be:
```
return cfg
```

(without the surrounding branch)?

Reply:
No. `cfg` can be `nil` without having an error (if config found for the given parameters on the thread) and we want to keep iterating through the loop in that case.